events/nats_config.go

```diff
@@ -0,0 +1,223 @@
+type NatsConfig struct {
+	URL             string
+	SubscribePrefix string
+	PublishPrefix   string
+	QueueGroup      string
+	Token           string
+	CredsFile       string
+	Source          string
+
+	ConnectTimeout           *time.Duration
+	ShutdownTimeout          *time.Duration
+	SubscriberFetchBatchSize *int
+	SubscriberFetchTimeout   *time.Duration
+	SubscriberFetchDelay     *time.Duration
+	SubscriberAckExplicit    *bool
+	SubscriberManualAck      *bool
+	SubscriberStartSequence  *uint64
+
+	Logger           *zap.SugaredLogger `mapstructure:"-"`
+	ConnectOptions   []nats.Option      `mapstructure:"-"`
+	JetStreamOptions []nats.JSOpt       `mapstructure:"-"`
+	SubscribeOptions []nats.SubOpt      `mapstructure:"-"`
+}
+
+// Configured checks whether the provider has been configured.
+func (c NatsConfig) Configured() bool {
+	return c.URL != "" || c.QueueGroup != ""
+}
+
+// Validate ensures the configuration is valid.
+func (c NatsConfig) Validate() error {
+	if c.Token != "" && c.CredsFile != "" {
+		return ErrNatsInvalidAuthConfiguration
+	}
+
+	return nil
+}
+
+// WithDefaults sets default values for the field unset.
+func (c NatsConfig) WithDefaults() NatsConfig {
+	if c.Logger == nil {
+		c.Logger = zap.NewNop().Sugar()
+	}
+
+	if c.SubscriberFetchBatchSize == nil {
+		c.SubscriberFetchBatchSize = &NATSDefaultSubscriberFetchBatchSize
+	}
+
+	if c.SubscriberFetchTimeout == nil {
+		c.SubscriberFetchTimeout = &NATSDefaultSubscriberFetchTimeout
+	}
+
+	if c.SubscriberFetchDelay == nil {
+		c.SubscriberFetchDelay = &NATSDefaultSubscriberFetchDelay
+	}
+
+	if c.SubscriberAckExplicit == nil || *c.SubscriberAckExplicit {
+		c.SubscribeOptions = append(c.SubscribeOptions, nats.AckExplicit())
+	}
+
+	if c.SubscriberManualAck == nil || *c.SubscriberManualAck {
+		c.SubscribeOptions = append(c.SubscribeOptions, nats.ManualAck())
+	}
+
+	if c.SubscriberStartSequence != nil {
+		c.SubscribeOptions = append(c.SubscribeOptions, nats.StartSequence(*c.SubscriberStartSequence))
+	}
+
+	if c.ShutdownTimeout == nil {
+		c.ShutdownTimeout = &NATSDefaultShutdownTimeout
+	}
+
+	if c.ConnectTimeout != nil {
+		c.ConnectOptions = append(c.ConnectOptions, nats.Timeout(*c.ConnectTimeout))
+	}
+
+	if c.Token != "" {
+		c.ConnectOptions = append(c.ConnectOptions, nats.Token(c.Token))
+	}
+
+	if c.CredsFile != "" {
+		c.ConnectOptions = append(c.ConnectOptions, nats.UserCredentials(c.CredsFile))
+	}
+
+	return c
+}
+
+// NatsOption defines a nats configuration option.
+type NatsOption func(c *NatsConfig) error
+
+// WithNatsLogger sets the logger for the nats connection.
+func WithNatsLogger(logger *zap.SugaredLogger) NatsOption {
+	return func(c *NatsConfig) error {
+		c.Logger = logger
+
+		return nil
+	}
+}
+
+// WithNatsConnectTimeout sets the timeout connecting to nats.
+func WithNatsConnectTimeout(timeout time.Duration) NatsOption {
+	return func(c *NatsConfig) error {
+		c.ConnectTimeout = &timeout
+
+		return nil
+	}
+}
```

Comment:
Maybe restating what I mentioned above but these functional options that map to config items risk introducing confusion where users may have multiple ways to configure something.